### ISSUE TITLE
Naming convention inspection false positive

### ISSUE THREAD

**Issue by matklad** [intellij-rust/intellij-rust#730]
```
match Some(()) {
    None => ()
}
```

The `None` here triggers inspection, while it shouldn't. The underlying problem is that we don't distinguish between name bindings and nullary enum variants in patterns. That is, for the following code

```
match Some(()) {
    None => (),
    Nane => (),
}
```

`None` is a constant, so it should not trigger the warning, while `Nane` is a variable name, and should be highlighted with red. I think we should just ignore variable names in patterns until we are able to properly distinguish the two cases. cc @alygin

**Comment by alygin** [comment]
It looks like the additional check `el.parent?.parent !is RustMatchPatElement` in the Variable Naming inspection solves the problem. Does it look reliable to you?

**Comment by matklad** [comment]
>Does it look reliable to you?

What about nested patterns?

```
match Some(Some(92)) {
    Some(None) => (),
    Some(Nane) => (),
    _ => (),
}
```

And this can happen in something like `if let` as well.

**Comment by alygin** [comment]
You're right. Then maybe use `el.parent?.parent is RustLetDeclElement` and don't try to cover such cases until the distinguishing is possible? It will filter out `match` and `if let` blocks.

**Comment by matklad** [comment]
I think it's better to avoid false positives, even if that means allowing some negatives. 

So I would go the opposite way and enable this inspection when we are sure it is a true positive. Let's check only identifiers of irrefutable patterns: 

```
// check these:
let Foo = 92;
let (Foo, Bar) = (1, 2);
fn foo(Foo: i32) { }

// and ignore these:
match x {
  Foo => { }
}

if let Foo = x { }
```

**Comment by matklad** [comment]
I am trying to say that to determine if the inspection is applicable, we need to see the context of the whole pattern, and not only of the identifier itself.

**Comment by alygin** [comment]
@matklad, could you, please, have a look at [the fix](https://github.com/intellij-rust/intellij-rust/pull/732)? There's no much of scrutiny there, but it seems doing the job.

While `match` blocks are out of question now, I think we could handle `if let`s correctly if put a bit more intelligence into the inspection. I'll get back to it shortly.
